### PR TITLE
DL-FIND constrained geometry optimization

### PR DESCRIPTION
**Pull request by akhilshajan**
I have added Dl-FInd feature for constrained geometry optimization.

```
# Example input for DL-FIND constraint integration in QUICK
HF BASIS=3-21G CONSTRAIN OPTIMIZE

C         -4.13498124  0.70342204  0.00000000
H         -3.53650966  -0.16516720  -0.17967886
H         -4.07172084  1.36057017  -0.84205371
H         -5.15338585  0.41046458  0.14808100
C         -3.62163902  1.42937832  1.25740497
H         -3.68306794  2.48719615  1.10858318
H         -4.22133413  1.15219316  2.09909028
H         -2.60384227  1.15314360  1.43819258

DIHEDRAL 2 1 5 6
```

The syntax for implementing constraints within QUICK closely mirrors established conventions in other QM packages: 

`FREEZEXY atom_no_1`
`DISTANCE atom_no_1 atom_no_2`
`ANGLE atom_no_1 atom_no_2 atom_no_3`
`DIHEDRAL atom_no_1 atom_no_2 atom_no_3 atom_no_4`

Commands such as DISTANCE, ANGLE, and DIHEDRAL enable straightforward specification of constraints, streamlining the optimization process. These constraints operate within the default internal coordinate system (ICOORD=3) of QUICK, simplifying the setup process for users familiar with similar packages.

The freeze atom commands (FREEZE, FREEZEX, FREEZEY, FREEZEZ, FREEZEXY, FREEZEXZ, FREEZEYZ) provide fine-grained control over molecular configurations, enhancing the precision and accuracy of simulations. The freeze atom functionality is compatible with Cartesian coordinate systems (ICOORD=0)of QUICK.

### CODE REVIEW

**Comment by Madu86** [comment]
@akhilshajan , I just tested this branch. Serial CPU and CUDA versions work fine; but not the MPI versions. Please check the attached zip file below. 
[ethane.zip](https://github.com/merzlab/QUICK/files/15194234/ethane.zip)

**Comment by Madu86** [comment]
Can you also add a test case to the test suite?

**Comment by akhilshajan** [comment]
Thank you Madu. I will check the error with MPI and add the test case with the test you tried out for constrained geometry optimization.

**Comment by akhilshajan** [comment]
Hi @Madu86, thanks with the help for MPI. I have tested with CUDA and CUDA-MPI. It is working perfectly now.
[ethane.zip](https://github.com/user-attachments/files/16150825/ethane.zip)

**Comment by Madu86** [comment]
Thanks @akhilshajan. We will test this soon.

**Comment by dacase** [comment]
I've had less success here.  I grabbed the code from `[akhilshajan:QUICK_Constrained_Optimization]`, built and tested it with the sample input above (and other sample inputs).  It never seems to recognize the CONSTRAIN keyword, just doing an unconstrained optimization.

I can be patient.  Perhaps I'll wait until this PR is merged in, then try again.  But just seeing if this rings any bells.  (I don't have time or expertise to dig into the code at this point.)

...thanks...dac

**Comment by Madu86** [comment]
Hi Dave, 

It looks like the quick executable is using a shared object file from another installation. Can you please run ldd on quick executable and check if this is the case? 

Madu

**Comment by dacase** [comment]
You are right to some extent.  I had to update my LD_LIBRARY_PATH  (doesn't seem to be done via `quick.rc`(?)  Now, `ldd` shows I am using `/usr/local/lib/libquick_cuda.so` instead of the one from Amber24.  But the results are the same: the CONSTRAIN keyword is apparently not recognized.  Probably I am still doing something wrong, but I don't see it.

**Comment by Madu86** [comment]
I still think this is a path issue. I have separate Amber24 and quick development (quick-const-opt) installations on my system.
This is what I compiled:
```
[madu@osprey quick-const-opt]$ git branch
* QUICK_Constrained_Optimization
[madu@osprey quick-const-opt]$ git log
commit 99797037c6309377b6f474bcc6bebbba78609cbf (HEAD -> QUICK_Constrained_Optimization, origin/QUICK_Constrained_Optimization)
Merge: 46ee0c15 1d6cf2d4
Author: Akhil Shajan <87541836+akhilshajan@users.noreply.github.com>
Date:   Thu May 2 13:52:55 2024 -0400

    Merge branch 'merzlab:master' into QUICK_Constrained_Optimization

```

Running ldd on quick.cuda from Amber24 installation gives me this:

```
[madu@osprey bin]$ ldd /home/madu/Apps/amber24/install/bin/quick.cuda
	linux-vdso.so.1 (0x00007ffd9afb4000)
	libquick_cuda.so => /home/madu/Apps/amber24/install//lib/libquick_cuda.so (0x00007fe27e400000)
```

And this is from QUICK development installation:

```
[madu@osprey bin]$ ldd /home/madu/Apps/quick-const-opt/install/bin/quick.cuda 
	linux-vdso.so.1 (0x00007ffce49ce000)
	libquick_cuda.so => /home/madu/Apps/quick-const-opt/install//lib/libquick_cuda.so (0x00007f37c6c00000)
```

Ethane test:

```
. /home/madu/Apps/quick-const-opt/install/quick.rc 
quick.cuda ethane.in
```

I have attached my input and output below. 
[ethane-rtx3080ti.tar.gz](https://github.com/user-attachments/files/17781549/ethane-rtx3080ti.tar.gz)

**Comment by dacase** [comment]
Ah...my fault.  I had built from the master branch, and not from Quick_Constrained_Optimization.  Thanks for your help; apologies for the noise. Everything now seems to work as I expected.

**Comment by Madu86** [comment]
No worries at all! Glad you figured it out and got everything working as expected. On Nov 15, 2024, at 6:56 PM, David A Case ***@***.***> wrote:﻿
Ah...my fault.  I had built from the master branch, and not from Quick_Constrained_Optimization.  Thanks for your help; apologies for the noise. Everything now seems to work as I expected.

**Comment by crtag** [comment]
A quick question - if this becomes a feature, would you please update the docs file too to reflect supported keywords etc?

**Comment by ohearnk** [comment]
@akhilshajan:  If you could add new tests for each of the new the DL-FIND optimization constraints, we would appreciate that.

**Comment by crtag** [comment]
@ohearnk A quick question - the only thing preventing the merge is missing tests, is that so? Logs from the last checks have already expired, so I'd like to know if there's anything else.

**Comment by ohearnk** [comment]
@crtag: There are still some MPI-related issues (with Intel oneAPI CCL) which need to be diagosed and corrected (see updated test artifacts with multiple test failures and a crash for different oneAPI versions).  There are also some MacOS build failures, though these are new and require MPI re-compilation with newer GCC gfortran in Homebrew to resolve  (i.e., it'll be fixed upstream eventually and are not really of concern right now).

Back to the oneAPI CCL issues -- there is likely an issue with the Fortran interface for DL-FIND (which I'm guessing hasn't been properly tested with MPI yet alone oneAPI CCL).  I've patched multiple compilation warnings / errors in the copy of DL-FIND included in QUICK with this MR already.  Ideally, @akhilshajan would debug this as he mainly worked on this feature.  Beyond these failures, @akhilshajan and/or @agoetz also need to add new tests to cover the new unconstrained optimization feature.

TL;DR -- this still needs work to fix localized MPI issues and needs more tests for the new features.  If this can be done before the QUICK-25.08 release later this year, then it'll be merged.

**Comment by crtag** [comment]
@ohearnk Thank you for the explanation! I will make a bold assumption that it can be tried in the current Ubuntu build, as the two failed checks may relate to missing tests, and I can benchmark it against the previous build with 24.x for sanity

**Comment by ohearnk** [comment]
@agoetz: This should be good to merge now once the additional constrained optimization tests are added.

**Review by agoetz — APPROVED** [review]
Constraints have been tested and remaining issues fixed.

Tested
- serial and MPI on macOS 15.6.1 with GCC 14.2.0, CLANG 19.1.7, OpenMPI 5.0.7.
- serial, MPI and CUDA on Almalinux 9.6 with GCC 11.5.0, OpenMPI 4.0.5, CUDA 12.4.1 on RTX2080 GPU.

This feature is ready to be merged.

**Comment by agoetz** [comment]
Tests have been added for
- distance, angle, and dihedral constraints.
- freeze atoms.
